[RFC] Utilize shared memory to deduplicate the network system-wide

Sopel97 opened the pull request:
This PR introduces a `SystemWideSharedConstant<T>` class that allows placing a content-addressable constant in shared-memory. It is deduplicated system-wide (for a single user). Compared to @Disservin's approach here https://github.com/Disservin/Stockfish/tree/wip-mmap-2 it maintains a single network structure and is less intrusive. Fully compatible and trivially integrated with our NUMA machinery (assuming shared memory allocation follows normal allocation rules with respect to NUMA node placement). Aside from performance improvements it also results in significant memory footprint reduction. This issue was explored and put into light by @AndyGrant [verbatim.pdf](https://github.com/user-attachments/files/21386224/verbatim.pdf).

Discord discussion: https://discord.com/channels/435943710472011776/813919248455827515/1397183730635772024

The current state of the code is that it works on Windows and is mostly complete, with a few ugly things to improve. Adding linux support should be minimal amount of work at this point.

What it involves:

1. The network structure can no longer use indirection, it needs to be trivial, i.e. the data needs to be stored in contiguous memory. This is required to make it work from shared memory without complex allocators and pointer->offset conversion.
2. Following the above, in some place dynamic allocations are required because the object is too large to fit on the stack. This is only for intermediate copies though.
3. Additional copies are created throughput during the initialization process, though this can be improved slightly, not much care has been taken to address this yet, old code just assumed the network object is cheap to move.
4. Writing the network to file now creates a copy of the feature transformer to avoid in-place modification and preserve const-correctness.
5. Proper hashing functions are added for network related structures, they are used to detrive content hash of the network for shared memory addressing.
6. A FixedString class that stores a string with fixed capacity without heap allocations, necessary for EvalFile as it's part of the network structure

---------------------------------

Test on windows 11 with large pages, 7800x3d, via script provided by @Disservin https://pastebin.com/89syg9mu:
<img width="3165" height="1764" alt="graph" src="https://github.com/user-attachments/assets/52897210-d584-4f18-bb9e-b3eb3d5277fe" />
```python compare.py --max-threads 16 --output out.txt --graph graph.png --iterations 2 "stockfish-shared.exe speedtest 1 16 5" "stockfish-master.exe speedtest 1 16 5" 
```

```==================================================
FINAL SUMMARY
==================================================

Command 1 Results:
Threads | Avg Nodes/s | Min | Max | Median | StdDev
-------------------------------------------------------
      1 |  1344021.00 | 1335100.00 | 1352942.00 | 1344021.00 | 12616.20
      2 |  1327394.00 | 1321839.00 | 1333741.00 | 1326998.00 | 5014.58
      4 |  1303861.75 | 1297324.00 | 1311121.00 | 1303747.00 | 4867.75
      8 |  1139116.00 | 1104799.00 | 1174991.00 | 1137885.50 | 20481.99
     16 |   770158.22 | 759475.00 | 778693.00 | 770336.00 | 4745.81

Command 2 Results:
Threads | Avg Nodes/s | Min | Max | Median | StdDev
-------------------------------------------------------
      1 |  1329460.00 | 1326731.00 | 1332189.00 | 1329460.00 | 3859.39
      2 |  1317041.00 | 1313630.00 | 1318601.00 | 1317966.50 | 2301.10
      4 |  1277356.75 | 1266281.00 | 1287642.00 | 1276741.00 | 7691.70
      8 |  1081963.94 | 1047183.00 | 1116268.00 | 1077491.00 | 19623.02
     16 |   690055.75 | 678430.00 | 707478.00 | 688052.50 | 7302.30```

More benchmarks from higher core machines are welcome.

Disservin commented:
nice, i can take a look at linux/macos impl again.. 
a) was the single core regression fixed now? (maybe run a longer speedup command)
b) the shm file needs an adapter depending on the os, shm_open which we'll probably use later only works up to `NAME_MAX` which is 255 chars, though on macos NAME_MAX was something really low like 20 iirc
c) was the issue fixed where it didn't work when multiple instances were spawned too quickly ? @vondele and I thought about this in the past and were maybe thinking about some sort of mutex which might be required ?
d) probably needs some numa testing later again, @vondele and I were in the past thinking about creating multiple shm files which are then mapped and each process in a numa domain would just point to a different mmapped file, this would get rid of most numa code except the affinity binding, no?
e) there were some quirks on linux to get huge pages to work with shm, because they had to be enabled for the underlying shm fs, not sure if windows really respects the system level large page support here.. ?

Disservin commented:
ah meaning, do we have a fallback in place if shared memory is not working ? i.e. in docker the default shm size is too small for out network

Sopel97 commented:
> I think Disservin's question higher up can maybe be clarified with an example.. if we have a 2 socket NUMA system, and start a fishtest instance on each of these sockets (with taskset properly matching the two sockets), will the SF instances on the one socket need to access the memory of the other socket for the network? Ideally, each numa domain has one copy of the network, and processes/threads always use the nearest copy.

There are 4 discriminators

1. Network content
2. Executable path
3. resolved `NumaPolicy`, so the string (i.e. "0-7:8-15") that means how processors are mapped to numa nodes
4. The NUMA node index

for each unique tuple of these discriminators one and only one network instance exists

So if there are 2 fishtest instances, each on a different socket, they will have different NumaPolicy, and will therefore NOT interfere with each other

While it won't actually detect that two nodes are physically the same I think a logical distinction is more desirable from user perspective.

Disservin commented:
> at least on my end it was due to large pages. Don't have other hardware to test on.

that's weird at least if I understand it correctly..
https://learn.microsoft.com/en-us/windows/win32/memory/creating-a-file-mapping-using-large-pages
the MapViewOfFile is missing `FILE_MAP_LARGE_PAGES` so it shouldn't really work ?

Disservin commented:
https://github.com/Disservin/Stockfish/commits/shared_memory_sopel/

here is my branch with the linux code, i removed the exe path discriminator for now but this is the branch where i get the slowdown, no MADV_HUGEPAGE flag for madvise/mmap worked for me..

Disservin commented:
okay so when I mount a tmpfs with `huge=always,size=2G` and move the file creation there the regression is gone.. and actually a speedup probably due to the even larger page?

```
sf_base =  1276171 +/-   9657 (95%)
sf_test =  1306195 +/-   8113 (95%)
diff    =    30023 +/-  11961 (95%)
speedup = 2.35263% +/- 0.937% (95%)
```

again really hacky and this requires root..

Sopel97 commented:
I'll try to rewrite this as an UCI option when I find some time

vondele commented:
there might still be a way around without uci options? Was there no memfd_create which allows for bypassing /dev/shm ?

Anyway, first local testing of this looks good. Still want to do some more testing.

Sopel97 commented:
> > there might still be a way around without uci options? Was there no memfd_create which allows for bypassing /dev/shm ?
> > Anyway, first local testing of this looks good. Still want to do some more testing.
> 
> When I told you about that I didn't realize that we'd have to share the file descriptor between the processes and there isn't a really good way to do this either other than writing it to a file I guess

two shared memories

Disservin commented:
mh.. i have to think about how to do this properly even.. fd aren't a global thing they are bound to the process from which they were created so other processes would have to read `/proc/pid/fd/fd`, but what if the initial process that created the file doesn't exist anymore.. ?


sometimes i had a positive effect with `MAP_SHARED | MAP_ANONYMOUS | MAP_POPULATE` in later benchmarks i didnt anymore.. idk

Sopel97 commented:
As I understand it these descriptors can actually be shared with other processes.

> •  The process that called memfd_create() could transfer the
               resulting file descriptor to the second process via a UNIX
               domain socket (see [unix(7)](https://man7.org/linux/man-pages/man7/unix.7.html) and [cmsg(3)](https://man7.org/linux/man-pages/man3/cmsg.3.html)).  The second
               process then maps the file using [mmap(2)](https://man7.org/linux/man-pages/man2/mmap.2.html).

so they could just be shared via shm shared memory

Disservin commented:
That only works with sockets, although there is pidfd_getfd to properly duplicate this, still doesn't help me in the case the pid process is no longer alive.. would require sharing all alive pids in a file and then taking one from the list I guess

See the second answer 

https://stackoverflow.com/questions/42319310/is-a-file-descriptor-local-to-its-process-or-global-on-unix

Disservin commented:
@Sopel97 in your implementation on windows I get the following output
```
Stockfish dev-20250723-78b403ee by the Stockfish developers (see AUTHORS file)
SystemWideSharedConstant total size: 146062212
SystemWideSharedConstant using normal pages...
initializing: Local\2063140025228914963$7883485429838667673$10995037620754966421
SystemWideSharedConstant total size: 146062212
SystemWideSharedConstant using normal pages...
initializing: Local\6149776985357895272$7883485429838667673$10995037620754966421
```

network gets init twice with same size but different hash ? i dont quite understand this

Sopel97 commented:
EvalFile has different content at first initialization I believe

Sopel97 commented:
Rebased on current master. Fixed compilation on Windows. Let's see what else needs to be fixed...

There is quite a lot of code that's system dependent but fairly generic, spread around misc/shm/other various files. Like, there's now GetLastErrorAsString in shm.h that I wanted to move to misc.h but it's not possible without polluting windows.h for pretty much every include. There will need to be some general rework in how system-dependent parts are handled because it's becoming annoying to maintain.

Disservin commented:
there's ashmem, but i think we should rather just check `_POSIX_SHARED_MEMORY_OBJECTS` and if not available use the fallback

Disservin commented:
> I just defined a dummy backend for android which always returns is_valid->false, as far as my understanding is, result in the the fallback backend.

yeah, as long as the system specific stuff is all properly guarded, could also just inherit the fallback backend instead but doesnt really matter

vondele commented on src/main.cpp:
```diff
@@ -17,28 +17,68 @@
+                     SIGSEGV, SIGTERM, SIGBUS,  SIGSYS, SIGXCPU, SIGXFSZ};
+
+    struct sigaction sa;
+    sa.sa_handler = [](int) { std::exit(1); };
```
```diff
-    sa.sa_handler = [](int) { std::exit(1); };
+    sa.sa_handler = [](int sig) { std::exit(128 + sig); };
```

vondele commented:
we still have some unexplained crashes:

https://tests.stockfishchess.org/actions?max_actions=1&action=&user=&text=shared_memory&before=1754592934.702885&run_id=

https://tests.stockfishchess.org/actions?max_actions=1&action=&user=&text=shared_memory&before=1754591010.967743&run_id=

the review comment should help extract the precise reason with the companion https://github.com/official-stockfish/fishtest/pull/2370

vondele commented:
looks fixed by https://github.com/vondele/Stockfish/commit/2d69ac6816a65c1e205e35e46f78f86e0840b1f1

vondele commented:
> So this is ready I guess?

probably nearly so, I'd like to test this once more without fleet active.

I wonder if there is a good way to test there are no similar race conditions on windows, since the linux one was a bit subtle, and required some extensive testing.

Maybe another review would be useful. Might try copilot on this one.

vondele commented on src/shm.h:
```diff
@@ -0,0 +1,571 @@
+
+namespace Stockfish {
+
+// Ideally use argv[0] instead maybe
```
is there anything that would speak against doing this?

Sopel97 replied:
argv[0] is not actually a path, it's a string used to invoke the program, so it can be the same for different paths. The comment needs to be removed or clarified the other way

vondele commented on src/shm.h:
```diff
@@ -0,0 +1,571 @@
+        }
+        else
+        {
+            std::cerr << "Failed to create shared memory backend: "
```
probably this should not be written to stderr, especially since it is not a hard error, and will be invisible to most people using a GUI or similar, but maybe use UCIEngine::print_info_string ? (a bit like we do in the verify networks)

Sopel97 replied:
Coupling it with UCI sounds bad, and doing it in a different way would be convoluted. I agree that it's not ideal, but potentially not that important to tackle at this point.

vondele replied:
I meant for the networks we pass a lambda doing
```
uci.cpp:    engine.set_on_verify_networks([](const auto& s) { print_info_string(s); });
uci.cpp:    engine.set_on_verify_networks([](const auto&) {});
```
which is roughly what I meant.

Sopel97 replied:
I think to achieve this we have to remove the printout here and instead check the status after loading the network. Will explore this.

vondele commented:
https://tests.stockfishchess.org/tests/view/68972e54f8a258623dda6bbf?show_task=39 

still an issue here, this is a worker run in docker, which by default has not sufficient space in /dev/shm the fallback should somehow catch this, but doesn't yet?

I guess @Disservin knows better where that could be caught.

vondele commented:
> I can take a look later today, also your fix patch looked like more of a hack, I'll try to take a look at that later too

I tend to agree with that... somehow I would expect one needs posix semaphores for what we're doing here.

Sopel97 commented:
I added std::launder on header_ptr_ to prevent optimizations from UB.

Ideally it should indeed use some form of an OS mutex like we do on windows.

vondele commented:
I was still looking a bit in the code to see if I could avoid the issue with the workers on docker that fail during the profile-build with sigbus. I tried to prepopulate the mmap'ed memory (and thus have it fail during mmap if there is not enough space), and while it is probably a good idea, it doesn't fix that particular issue https://tests.stockfishchess.org/tests/view/689752a5f8a258623dda6c19 . MAP_POPULATE is linux only, so might not cover all cases. 
I wonder if we should go a step further with mlock.. but idk.

Disservin commented:
this works for me

```diff
diff --git a/src/shm_linux.h b/src/shm_linux.h
index 50599aacd..cce6b0a2c 100644
--- a/src/shm_linux.h
+++ b/src/shm_linux.h
@@ -296,6 +296,13 @@ class SharedMemory: public detail::SharedMemoryBase {
             return false;
         }

+        int r = posix_fallocate(fd_, 0, total_size_);
+        if (r != 0)
+        {
+            sem_post(sem_);
+            return false;
+        }
+
         // Map the memory
         auto flags  = PROT_READ | PROT_WRITE;
         mapped_ptr_ = mmap(nullptr, total_size_, flags, MAP_SHARED, fd_, 0);
```

vondele commented:
pushed that change to my branch, and scheduled testing.

vondele commented:
BTW, on linux this gives a good overview of how the memory is being handled:
`pmap  -XX -p 1378392`
```
$ pmap  -XX -p 1378392
1378392:   ./stockfish.sopel
         Address Perm   Offset Device    Inode   Size KernelPageSize MMUPageSize    Rss   Pss Shared_Clean Shared_Dirty Private_Clean Private_Dirty Referenced Anonymous LazyFree AnonHugePages ShmemPmdMapped FilePmdMapped Shared_Hugetlb Private_Hugetlb Swap SwapPss Locked THPeligible                 VmFlags Mapping
    55675024b000 r--p 00000000 103:01  9309607     24              4           4     24     0           24            0             0             0         24         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /home/vondele/chess/vondele/matetrack/stockfish.sopel
    556750251000 r-xp 00006000 103:01  9309607    284              4           4    284     9          284            0             0             0        284         0        0             0              0             0              0               0    0       0      0           0       rd ex mr mw me sd /home/vondele/chess/vondele/matetrack/stockfish.sopel
    556750298000 r--p 0004d000 103:01  9309607  76624              4           4  76624  2554        76624            0             0             0      76624         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /home/vondele/chess/vondele/matetrack/stockfish.sopel
    556754d6d000 r--p 04b21000 103:01  9309607      4              4           4      4     4            0            0             0             4          4         4        0             0              0             0              0               0    0       0      0           0       rd mr mw me ac sd /home/vondele/chess/vondele/matetrack/stockfish.sopel
    556754d6e000 rw-p 04b22000 103:01  9309607      4              4           4      4     4            0            0             0             4          4         4        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd /home/vondele/chess/vondele/matetrack/stockfish.sopel
    556754d6f000 rw-p 00000000  00:00        0   1112              4           4   1108  1108            0            0             0          1108       1108      1108        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    556755797000 rw-p 00000000  00:00        0    264              4           4    108   108            0            0             0           108        108       108        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd [heap]
    7f545abf1000 rw-s 00000000  00:1a     2276 142640              4           4 138992  4778            0       138992             0             0     138992         0        0             0              0             0              0               0    0       0      0           0 rd wr sh mr mw me ms sd /dev/shm/cb3e3f4f978b31ef
    7f5464000000 rw-p 00000000  00:00        0  15860              4           4  15860 15860            0            0             0         15860      15860     15860        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me nr sd 
    7f5464f7d000 ---p 00000000  00:00        0  49676              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0          mr mw me nr sd 
    7f546a887000 rw-p 00000000  00:00        0   1508              4           4      8     8            0            0             0             8          8         8        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    7f546aa00000 rw-p 00000000  00:00        0  16384              4           4  16384 16384            0            0             0         16384      16384     16384        0         16384              0             0              0               0    0       0      0           1 rd wr mr mw me ac sd hg 
    7f546ba00000 rw-p 00000000  00:00        0    544              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    7f546ba88000 ---p 00000000  00:00        0      4              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0             mr mw me sd 
    7f546ba89000 rw-p 00000000  00:00        0   8220              4           4    236   236            0            0             0           236        236       236        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    7f546c290000 r--p 00000000 103:01 10489065    160              4           4    160     0          160            0             0             0        160         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libc.so.6
    7f546c2b8000 r-xp 00028000 103:01 10489065   1620              4           4   1120     5         1120            0             0             0       1120         0        0             0              0             0              0               0    0       0      0           0       rd ex mr mw me sd /usr/lib/x86_64-linux-gnu/libc.so.6
    7f546c44d000 r--p 001bd000 103:01 10489065    352              4           4    192     0          192            0             0             0        192         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libc.so.6
    7f546c4a5000 ---p 00215000 103:01 10489065      4              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0             mr mw me sd /usr/lib/x86_64-linux-gnu/libc.so.6
    7f546c4a6000 r--p 00215000 103:01 10489065     16              4           4     16    16            0            0             0            16         16        16        0             0              0             0              0               0    0       0      0           0       rd mr mw me ac sd /usr/lib/x86_64-linux-gnu/libc.so.6
    7f546c4aa000 rw-p 00219000 103:01 10489065      8              4           4      8     8            0            0             0             8          8         8        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd /usr/lib/x86_64-linux-gnu/libc.so.6
    7f546c4ac000 rw-p 00000000  00:00        0     52              4           4     20    20            0            0             0            20         20        20        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    7f546c4b9000 r--p 00000000 103:01 10532070     12              4           4     12     0           12            0             0             0         12         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
    7f546c4bc000 r-xp 00003000 103:01 10532070    108              4           4     64     1           64            0             0             0         64         0        0             0              0             0              0               0    0       0      0           0       rd ex mr mw me sd /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
    7f546c4d7000 r--p 0001e000 103:01 10532070     16              4           4     16     0           16            0             0             0         16         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
    7f546c4db000 r--p 00021000 103:01 10532070      4              4           4      4     4            0            0             0             4          4         4        0             0              0             0              0               0    0       0      0           0       rd mr mw me ac sd /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
    7f546c4dc000 rw-p 00022000 103:01 10532070      4              4           4      4     4            0            0             0             4          4         4        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
    7f546c4dd000 r--p 00000000 103:01 10489190     56              4           4     56     0           56            0             0             0         56         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libm.so.6
    7f546c4eb000 r-xp 0000e000 103:01 10489190    496              4           4    320     3          320            0             0             0        320         0        0             0              0             0              0               0    0       0      0           0       rd ex mr mw me sd /usr/lib/x86_64-linux-gnu/libm.so.6
    7f546c567000 r--p 0008a000 103:01 10489190    364              4           4    128     3          128            0             0             0        128         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libm.so.6
    7f546c5c2000 r--p 000e4000 103:01 10489190      4              4           4      4     4            0            0             0             4          4         4        0             0              0             0              0               0    0       0      0           0       rd mr mw me ac sd /usr/lib/x86_64-linux-gnu/libm.so.6
    7f546c5c3000 rw-p 000e5000 103:01 10489190      4              4           4      4     4            0            0             0             4          4         4        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd /usr/lib/x86_64-linux-gnu/libm.so.6
    7f546c5c4000 r--p 00000000 103:01 10532071    624              4           4    624    19          624            0             0             0        624         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.32
    7f546c660000 r-xp 0009c000 103:01 10532071   1220              4           4    712    22          712            0             0             0        712         0        0             0              0             0              0               0    0       0      0           0       rd ex mr mw me sd /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.32
    7f546c791000 r--p 001cd000 103:01 10532071    564              4           4    124     5          124            0             0             0        124         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.32
    7f546c81e000 ---p 0025a000 103:01 10532071      4              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0             mr mw me sd /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.32
    7f546c81f000 r--p 0025a000 103:01 10532071     44              4           4     44    44            0            0             0            44         44        44        0             0              0             0              0               0    0       0      0           0       rd mr mw me ac sd /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.32
    7f546c82a000 rw-p 00265000 103:01 10532071     12              4           4     12    12            0            0             0            12         12        12        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.32
    7f546c82d000 rw-p 00000000  00:00        0     16              4           4     12    12            0            0             0            12         12        12        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    7f546c855000 rw-p 00000000  00:00        0      8              4           4      8     8            0            0             0             8          8         8        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd 
    7f546c857000 r--p 00000000 103:01 10485864      8              4           4      8     0            8            0             0             0          8         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
    7f546c859000 r-xp 00002000 103:01 10485864    168              4           4    168     0          168            0             0             0        168         0        0             0              0             0              0               0    0       0      0           0       rd ex mr mw me sd /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
    7f546c883000 r--p 0002c000 103:01 10485864     44              4           4     44     0           44            0             0             0         44         0        0             0              0             0              0               0    0       0      0           0          rd mr mw me sd /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
    7f546c88e000 rw-s 00000000  00:1a     2275      4              4           4      4     0            0            4             0             0          4         0        0             0              0             0              0               0    0       0      0           0 rd wr sh mr mw me ms sd /dev/shm/sem.cb3e3f4f978b31ef_mutex
    7f546c88f000 r--p 00037000 103:01 10485864      8              4           4      8     8            0            0             0             8          8         8        0             0              0             0              0               0    0       0      0           0       rd mr mw me ac sd /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
    7f546c891000 rw-p 00039000 103:01 10485864      8              4           4      8     8            0            0             0             8          8         8        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me ac sd /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
    7fff47ec2000 rw-p 00000000  00:00        0    132              4           4     92    92            0            0             0            92         92        92        0             0              0             0              0               0    0       0      0           0    rd wr mr mw me gd ac [stack]
    7fff47f44000 r--p 00000000  00:00        0     16              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0    rd mr pf io de dd sd [vvar]
    7fff47f48000 r-xp 00000000  00:00        0      8              4           4      4     0            4            0             0             0          4         0        0             0              0             0              0               0    0       0      0           0    rd ex mr mw me de sd [vdso]
ffffffffff600000 --xp 00000000  00:00        0      4              4           4      0     0            0            0             0             0          0         0        0             0              0             0              0               0    0       0      0           0                      ex [vsyscall]
                                               ====== ============== =========== ====== ===== ============ ============ ============= ============= ========== ========= ======== ============= ============== ============= ============== =============== ==== ======= ====== =========== 
                                               319324            200         200 253636 41355        80684       138996             0         33956     253636     33956        0         16384              0             0              0               0    0       0      0           1 KB 

```

Sopel97 commented:
That makes sense, it essentially allows higher reuse in L3, but is still limited by it. Might be problematic to calibrate to be fully consistent across machines.

Is there actually a measurable slowdown on 1 thread?

AndyGrant commented:
https://discord.com/channels/435943710472011776/735707599353151579/1396504227592798241

I've said as much before on the measurements being inflatable. A cheap, partial solution, is to just randomize the order of the benched positions, and make sure there is no functional diff with the TT age. But you still get some extra sharing you would not normally get. I see something like a > 100% gain in a bench, 30-60% gain in a randomized bench, 20-40% gain in gameplay.

Its important to note that for gameplay, you have to measure the speedup in terms of (master vs master) vs (patch vs patch), not (master vs patch). This patch speeds up all other engines that are playing. I posted this data once before:


> Verbatim Torch, compared to normal Torch, gains +11.51% speed against SF. But SF gains +7.10%.
> Verbatim Torch self-play, compared to normal Torch self-play, gains +39.75%. 
> Verbatim Torch, compared to normal Torch, gains +12.51% speed against Ethereal. But Ethereal gains +2.05%.
> 
> We see SF speeds up as well, because when Torch reduces its memory footprint, SF benefits.
> We see Ethereal speed up as well, but much less. Likely since Ethereal's memory footprint is very small compared to SF/Torch.
> 
> normal-Torch is 3.6% faster when playing itself, than against Stockfish. SF has a large footprint that hurts us.
> normal-Torch is 36% faster when playing against Ethereal, than against itself

vondele commented:
> I've said as much before on the measurements being inflatable. 

right, so here, and in several more instances, you are sharing the graph of inflated numbers, knowing they are a measurement artifact?

https://discord.com/channels/435943710472011776/813919248455827515/1396393406086778982

AndyGrant commented:
> > I've said as much before on the measurements being inflatable.
> 
> right, so here, and in several more instances, you are sharing the graph of inflated numbers, knowing they are a measurement artifact?
> 
> https://discord.com/channels/435943710472011776/813919248455827515/1396393406086778982
> 
> Edit: and this one is plain wrong https://discord.com/channels/435943710472011776/882956631514689597/1392309734715166832

I provided the above information about speedups to help you guys have a productive conversation about what to do for Fishtest. Not every time I post that graphic do I have the time and energy to do a deep dive on the context and the nuance. However, when going into details in conversations with Disservin and Sopel, I do just that. Both of them are aware of the fickle nature of the measurements. Disservin from the link I posted -- Sopel from our shared experience when testing the original NUMA network-duplication patches on CCC.

I don't have any interest in continuing to play out you not liking me. Take that somewhere else instead of detracting from the productive work being done here.

vondele commented:
I have just pointed out that the graphs you have posted were based on flawed measurements and the statement of 2x throughput  increase on fishtest was wrong. Even on hardware where this change has most effect, the measured result is below that.

I'm not exactly sure what the format of your pgns is, but I get more like 40% ... 
```
$ grep -o "{[0-9a-z\/\.\-\ ]*}" games-master.pgn | sed "s/{//" | sed "s/}//g" | awk '{t=t+$3;s=s+$4}END{print 1000*s/t}'
570478
$ grep -o "{[0-9a-z\/\.\-\ ]*}" games-verbatim.pgn | sed "s/{//" | sed "s/}//g" | awk '{t=t+$3;s=s+$4}END{print 1000*s/t}'
784949

--> 1.37
```
I might be wrong though, maybe you can post what you used to analyze it.

For my own tests on similar hardware, I do get 42% and somewhat beefier hardware 37%:
 
```

for bin in sopel master
do

echo " ===== $bin ===== "

start=$(date +%s)
rm -f ${bin}.pgn

./fastchess -concurrency 32 -rounds 320 -games 2 -repeat\
            -srand 42 -openings file=UHO_Lichess_4852_v1.epd format=epd order=random\
            -engine name=${bin}1 cmd=./stockfish.${bin}\
            -engine name=${bin}2 cmd=./stockfish.${bin}\
            -each proto=uci option.Threads=1 option.Hash=16 tc=10+0.1\
            -pgnout file=${bin}.pgn nodes=true nps=true >& out.${bin}

nodes=$(grep -o 'n=[0-9]*' ${bin}.pgn  | cut -d= -f2- | awk '{s=s+$1}END{print s}')
end=$(date +%s)

echo "$((end - start)) seconds for $nodes nodes"
echo $start $end $nodes | awk '{print "nps: ", $3/($2 - $1)}'

done
```

```
===== sopel =====
709 seconds for 5309140368 nodes
nps:  7.48821e+06
 ===== master =====
711 seconds for 3740186907 nodes
nps:  5.26046e+06

--> 1.42
```
```
 ===== sopel =====
353 seconds for 19509839936 nodes
nps:  5.52687e+07
 ===== master =====
348 seconds for 13978842506 nodes
nps:  4.01691e+07

--> 1.37590
```

robertnurnberg commented:
So I played with the this a bit. I am trying to improve on the above script (on a mobile!), and hence often interrupted the fastchess tournaments with ctrl-c.

Sadly, I have now reached a stage where the patch is unresponsive to Uci commands on startup. So each new tournament fails. 

Also tried compiler command from cli. This now enters interactive shell rather than printing info and quit. Also quit command no longer works.

```
> ./stockfish.sopel compiler
Stockfish dev-20250809-8274fc5c by the Stockfish developers (see AUTHORS file)
quit
^C
```

Any idea how I can get out of this? Without rebooting the machine. Of course, something to fix before merge as well, I think.

vondele commented:
yeah, you probably have a lock file that can't get cleaned up because the process that owns it dies. The workaround right now is to remove files in `/dev/shm` that you own. 

This is a problem with posix semaphores. Probably requires a change to use pthread_mutexattr_setrobust and no longer posix semaphores.

vondele commented:
> do we follow this for crapos? https://chromium.googlesource.com/native_client/src/native_client/+/d6d53e83fffcc64b6e0d52acc31c63e0b5a60186/src/shared/platform/osx/nacl_semaphore.c#46

not fully, but hardcoding SEM_NAME_LEN would fix the issue we see in CI. 

Having said that, to avoid the issue @robertnurnberg (and I at an earlier stage) experienced, we would have to drop posix semaphores...

Sopel97 commented:
right, I missed it, sorry. So file locking?

AndyGrant commented:
> `This issue was explored and put into light by @AndyGrant [verbatim.pdf](https://github.com/user-attachments/files/21386224/verbatim.pdf).`
> 
> Can I get credit somewhere for the initial identification of this problem: [official-stockfish/fishtest#2077](https://github.com/official-stockfish/fishtest/issues/2077) and the subsequent demonstration of the solution through the first impl of net sharing in a chess engine: [official-monty/Monty#62](https://github.com/official-monty/Monty/pull/62) (as well as discovering the behaviour of elo gain in multi process SPRT conditions etc.). AGEs work built on top of this

My exploration is indeed predicated on Viren's observations.

vondele commented:
> right, I missed it, sorry. So file locking?

well in /dev/shm... this ithe example chatgpt cooked up... though I didn't test it...

```
#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <pthread.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>
#include <time.h>

#define SHM_NAME "/portable_mutex_demo"
#define MAGIC_INIT 0xdeadbeef

struct shared_area {
    pthread_mutex_t mutex;
    int initialized;
};

/* Helper: fatal error */
static void fatal(const char *msg) {
    perror(msg);
    exit(EXIT_FAILURE);
}

/* Helper: portable lock with owner-death handling or timeout */
static int portable_mutex_lock(pthread_mutex_t *mtx, int timeout_sec) {
#ifdef PTHREAD_MUTEX_ROBUST
    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    ts.tv_sec += timeout_sec;

    int rc = pthread_mutex_timedlock(mtx, &ts);
    if (rc == 0) {
        return 0; // success
    } else if (rc == EOWNERDEAD) {
        // Owner died — repair shared state here if needed
        fprintf(stderr, "[WARN] Owner died, marking mutex consistent\n");
        pthread_mutex_consistent(mtx);
        return 0;
    } else {
        return rc; // ETIMEDOUT or other error
    }
#else
    // No robust support — emulate with timedlock only
    struct timespec ts;
    clock_gettime(CLOCK_REALTIME, &ts);
    ts.tv_sec += timeout_sec;
    int rc = pthread_mutex_timedlock(mtx, &ts);
    return rc;
#endif
}

int main(void) {
    int fd;
    int creator = 0;

    fd = shm_open(SHM_NAME, O_RDWR | O_CREAT | O_EXCL, 0660);
    if (fd >= 0) {
        creator = 1;
        if (ftruncate(fd, sizeof(struct shared_area)) == -1)
            fatal("ftruncate");
    } else if (errno == EEXIST) {
        fd = shm_open(SHM_NAME, O_RDWR, 0);
        if (fd == -1) fatal("shm_open");
    } else {
        fatal("shm_open");
    }

    struct shared_area *area = mmap(NULL, sizeof(*area),
                                    PROT_READ | PROT_WRITE,
                                    MAP_SHARED, fd, 0);
    if (area == MAP_FAILED) fatal("mmap");
    close(fd);

    if (creator) {
        pthread_mutexattr_t attr;
        pthread_mutexattr_init(&attr);
        pthread_mutexattr_setpshared(&attr, PTHREAD_PROCESS_SHARED);

#ifdef PTHREAD_MUTEX_ROBUST
        if (pthread_mutexattr_setrobust(&attr, PTHREAD_MUTEX_ROBUST) != 0)
            fprintf(stderr, "[WARN] robust mutex not supported here\n");
#endif

        pthread_mutex_init(&area->mutex, &attr);
        pthread_mutexattr_destroy(&attr);

        area->initialized = MAGIC_INIT;
        fprintf(stderr, "[INFO] Creator initialized mutex\n");
    } else {
        while (area->initialized != MAGIC_INIT) {
            usleep(1000);
        }
    }

    // Try to lock
    int rc = portable_mutex_lock(&area->mutex, 5); // 5s timeout
    if (rc == 0) {
        fprintf(stderr, "[INFO] Got lock, critical section...\n");
        sleep(2);
        pthread_mutex_unlock(&area->mutex);
    } else if (rc == ETIMEDOUT) {
        fprintf(stderr, "[ERROR] Timeout acquiring lock\n");
    } else {
        fprintf(stderr, "[ERROR] Lock failed: %s\n", strerror(rc));
    }

    if (creator) shm_unlink(SHM_NAME);
    return 0;
}
```

vondele commented:
thanks for the numbers and the script (I fixed a small pasto in the script).

vondele commented:
so on a 16c/32t CPU:
```
 ===== concurrency: 1 =====
 ===== sopel =====
702 seconds for 811291657 nodes
nps:  1.15569e+06
 ===== master =====
699 seconds for 802189922 nodes
nps:  1.14763e+06
concurrency 1 speedup: 1.00702
 ===== concurrency: 2 =====
 ===== sopel =====
738 seconds for 1637764864 nodes
nps:  2.21919e+06
 ===== master =====
710 seconds for 1656871353 nodes
nps:  2.33362e+06
concurrency 2 speedup: 0.95097
 ===== concurrency: 4 =====
 ===== sopel =====
716 seconds for 2991088304 nodes
nps:  4.17750e+06
 ===== master =====
694 seconds for 2861611591 nodes
nps:  4.12336e+06
concurrency 4 speedup: 1.01313
 ===== concurrency: 8 =====
 ===== sopel =====
717 seconds for 5163964657 nodes
nps:  7.20218e+06
 ===== master =====
706 seconds for 4916893110 nodes
nps:  6.96444e+06
concurrency 8 speedup: 1.03414
 ===== concurrency: 16 =====
 ===== sopel =====
720 seconds for 6943171134 nodes
nps:  9.64329e+06
 ===== master =====
711 seconds for 5897928439 nodes
nps:  8.29526e+06
concurrency 16 speedup: 1.16251
 ===== concurrency: 32 =====
 ===== sopel =====
710 seconds for 5422687573 nodes
nps:  7.63759e+06
 ===== master =====
721 seconds for 3871704407 nodes
nps:  5.36991e+06
concurrency 32 speedup: 1.42229
```
<img width="1596" height="1106" alt="image" src="https://github.com/user-attachments/assets/957e0bf7-5756-46ff-8379-0aca7686f834" />
<img width="1610" height="1114" alt="image" src="https://github.com/user-attachments/assets/a48402bf-9094-4839-8614-94d7b93c2115" />

robertnurnberg commented:
I have updated https://github.com/official-stockfish/Stockfish/pull/6173#issuecomment-3187892449 with some more numbers. 

I would be very curious to know if others with AMD Ryzen chips see similarly underwhelming numbers.

Sopel97 commented:
Alright, want to push this through to be done with it. MacOS needs a crude fix for CI, what's the state of the rest? in particular

> we would have to drop posix semaphores...

?

I really don't want to touch linux support myself if possible

vondele commented:
Yes, would need further work to not let this go stale.
* Linux semaphore work to fix potential deadlocks
* MacOS CI
* Understand cases of poor performance (some workers had opposite effect in last fishtest runs).

Disservin commented:
i couldn't find a header where this is defined on macos but tried creating semaphores with increasing name length and it failed when the name is longer than 30 (31 including null terminator) chars,
so this just needs to be updated for apple to be
`#define SF_MAX_SEM_NAME_LEN 31`

and 
```c++
if (shm_name.size() > NAME_MAX)
{
    shm_name = shm_name.substr(0, NAME_MAX);
}
```

needs to be changed to
```c++
if (shm_name.size() > NAME_MAX - 1)
{
    shm_name = shm_name.substr(0, NAME_MAX - 1);
}
```

Sopel97 commented:
Addressed https://github.com/official-stockfish/Stockfish/pull/6173#discussion_r2264675685, now the network allocation status is retrived by the engine after network is verified, the information is printed via on_verify_networks, through UCI.

```C:\dev\stockfish-master\src>stockfish.exe
Stockfish dev-20250901-nogit by the Stockfish developers (see AUTHORS file)
setoption name NumaPolicy value 1:2:3
info string Available processors: 1:2:3
info string Using 1 thread with NUMA node thread binding: 1/1:0/1:0/1
go nodes 1
info string Available processors: 1:2:3
info string Using 1 thread with NUMA node thread binding: 1/1:0/1:0/1
info string NNUE evaluation using nn-1c0000000000.nnue (133MiB, (22528, 3072, 15, 32, 1))
info string NNUE evaluation using nn-37f18f62d772.nnue (6MiB, (22528, 128, 15, 32, 1))
info string Network replica 1: Shared memory.
info string Network replica 2: No allocation.
info string Network replica 3: No allocation.
info depth 1 seldepth 2 multipv 1 score cp 17 nodes 20 nps 6666 hashfull 0 tbhits 0 time 3 pv e2e4
info depth 2 seldepth 2 multipv 1 score cp 17 nodes 21 nps 7000 hashfull 0 tbhits 0 time 3 pv e2e4
bestmove e2e4
setoption name threads value 3
info string Using 3 threads with NUMA node thread binding: 1/1:1/1:1/1
go nodes 1
info string Available processors: 1:2:3
info string Using 3 threads with NUMA node thread binding: 1/1:1/1:1/1
info string NNUE evaluation using nn-1c0000000000.nnue (133MiB, (22528, 3072, 15, 32, 1))
info string NNUE evaluation using nn-37f18f62d772.nnue (6MiB, (22528, 128, 15, 32, 1))
info string Network replica 1: Shared memory.
info string Network replica 2: Shared memory.
info string Network replica 3: Shared memory.
info depth 1 seldepth 2 multipv 1 score cp 17 nodes 20 nps 6666 hashfull 0 tbhits 0 time 3 pv e2e4
info depth 2 seldepth 2 multipv 1 score cp 17 nodes 447 nps 149000 hashfull 0 tbhits 0 time 3 pv e2e4
bestmove e2e4 ponder d7d5```

Torom commented:
[master](https://github.com/official-stockfish/Stockfish/commit/75edbee01e6f8cb53a2555499192ccaddb883577) vs [shared_memory_refactor_rebased](https://github.com/vondele/Stockfish/commit/cef1070fa9877c0cfd95863b67308aabbdd5baaf)
Oracle Cloud 24 threads of AMD EPYC 9J14
```
===== concurrency: 1 =====
 ===== patch =====
685 seconds for 918085647 nodes
nps:  1.34027e+06
 ===== master =====
694 seconds for 875983537 nodes
nps:  1.26222e+06
concurrency 1 speedup: 1.06183
 ===== concurrency: 2 =====
 ===== patch =====
683 seconds for 1762991122 nodes
nps:  2.58125e+06
 ===== master =====
690 seconds for 1752587368 nodes
nps:  2.53998e+06
concurrency 2 speedup: 1.01625
 ===== concurrency: 4 =====
 ===== patch =====
688 seconds for 3308135017 nodes
nps:  4.80834e+06
 ===== master =====
706 seconds for 3484157343 nodes
nps:  4.93507e+06
concurrency 4 speedup: 0.97432
 ===== concurrency: 8 =====
 ===== patch =====
701 seconds for 6379681561 nodes
nps:  9.10083e+06
 ===== master =====
706 seconds for 6380428151 nodes
nps:  9.03743e+06
concurrency 8 speedup: 1.00701
 ===== concurrency: 16 =====
 ===== patch =====
712 seconds for 9922401851 nodes
nps:  1.39360e+07
 ===== master =====
705 seconds for 9155995212 nodes
nps:  1.29872e+07
concurrency 16 speedup: 1.07305
```

Sopel97 commented on src/shm.h:
```diff
@@ -0,0 +1,646 @@
+        shm1(shm::create_shared<T>(shm_name, value)) {}
+
+    void* get() const {
+        const T* ptr = &shm1->get();
```
same guarantees as dereferencing a std::nullopt, is_valid should be checked first

anematode commented on src/nnue/network.cpp:
```diff
@@ -185,7 +146,7 @@ bool Network<Arch, Transformer>::save(const std::optional<std::string>& filename
     else
     {
-        if (evalFile.current != evalFile.defaultName)
+        if (std::string(evalFile.current) != std::string(evalFile.defaultName))
```
we define operator != on `FixedString` above, are these conversions still necessary?

Sopel97 replied:
it's a bit of a mess overall, the operator != as is defined is not great either, I'd honestly rather leave it at this point and revise all std::string usage across the codebase, favoring std::string_view and explicit conversions

anematode commented on src/nnue/network.h:
```diff
@@ -49,6 +51,9 @@ enum class EmbeddedNNUEType {
+// The network must be a trivial type, i.e. the memory must be in-line.
+// This is required to allow sharing the network via shared memory, as
+// there is no way to run destructors.
 template<typename Arch, typename Transformer>
```
Should we add a `static_assert(std::is_trivially_destructible_v<...>)` somewhere? Is that not possible?

vondele commented:
> Threadripper PRO 3995WX

we did look a little closer in this one, and found that CPU frequency throttling is happening. Interestingly, this happens for this branch clearly more than master. For now, we can assume that this is the HW, IMO.

That probably resolves the timings in https://github.com/official-stockfish/Stockfish/pull/6173#issuecomment-3187892449 and

Torom commented:
[master](https://github.com/official-stockfish/Stockfish/commit/bc9f08731f10896a306bcc34e30e5606087af79a) vs [shared_memory](https://github.com/Sopel97/Stockfish/commit/f0782933e6ced487d601be70bb1d5b3fb9714400)
Ryzen 9 9550X3D 32 threads
```
 ===== concurrency: 1 =====
 ===== shared_memory =====
733 seconds for 2015985614 nodes
nps:  2.75032e+06
 ===== master =====
759 seconds for 1897038404 nodes
nps:  2.49939e+06
concurrency 1 speedup: 1.10040
 ===== concurrency: 2 =====
 ===== shared_memory =====
676 seconds for 3238008717 nodes
nps:  4.78995e+06
 ===== master =====
694 seconds for 3162674326 nodes
nps:  4.55717e+06
concurrency 2 speedup: 1.05108
 ===== concurrency: 4 =====
 ===== shared_memory =====
698 seconds for 6012831940 nodes
nps:  8.61437e+06
 ===== master =====
700 seconds for 5810253794 nodes
nps:  8.30036e+06
concurrency 4 speedup: 1.03783
 ===== concurrency: 8 =====
 ===== shared_memory =====
713 seconds for 10715691220 nodes
nps:  1.50290e+07
 ===== master =====
698 seconds for 9051360485 nodes
nps:  1.29676e+07
concurrency 8 speedup: 1.15897
 ===== concurrency: 16 =====
 ===== shared_memory =====
736 seconds for 22005774396 nodes
nps:  2.98991e+07
 ===== master =====
703 seconds for 18257632945 nodes
nps:  2.59710e+07
concurrency 16 speedup: 1.15125
 ===== concurrency: 32 =====
 ===== shared_memory =====
716 seconds for 23753756934 nodes
nps:  3.31756e+07
 ===== master =====
717 seconds for 16033856940 nodes
nps:  2.23624e+07
concurrency 32 speedup: 1.48354
```
1 thread concurrency 1
```
Result of 200 runs
==================
base (...kfish.master) =    2314669  +/- 4178
test (...hared_memory) =    2299442  +/- 4349
diff                   =     -15228  +/- 2181

speedup        = -0.0066
P(speedup > 0) =  0.0000
```
32 threads concurrency 1
```
sf_base = 42634730 +/-  59598 (95%)
sf_test = 42616511 +/-  53826 (95%)
diff    =   -18219 +/-  74982 (95%)
speedup = -0.04273% +/- 0.176% (95%)
```
